source/features/useful-not-found-page.tsx

```diff
@@ -36,55 +38,112 @@ function parseCurrentURL(): string[] {
 	return parts;
 }
 
-// If the resource was deleted, link to the commit history
-async function addCommitHistoryLink(bar: Element): Promise<void> {
-	const parts = parseCurrentURL();
-	parts[2] = 'commits';
-	const url = '/' + parts.join('/');
-	if (await is404(location.origin + url)) {
+async function getLatestChangeToFile(): Promise<Record<string, any> | void> {
+	const url = new GitHubURL(location.href);
+	const currentBranch = url.branch;
+	const {filePath} = url;
+	if (!currentBranch || !filePath) {
 		return;
 	}
 
-	bar.after(
-		<p className="container mt-4 text-center">
-			See also the {getType()}’s <a href={url}>commit history</a>.
-		</p>,
-	);
+	const commitsResponseObject = await api.v4(`
+		repository() {
+			ref(qualifiedName: "${currentBranch}") {
+				target {
+					... on Commit {
+						history(first: 1, path: "${filePath}") {
+							nodes {
+								oid
+							}
+						}
+					}
+				}
+			}
+		}
+	`);
+	const commits =
+		commitsResponseObject.repository.ref.target.history.nodes[0];
+	if (!commits[0]) {
+		return;
+	}
+
+	// API v4 doesn't support retrieving a list of changed files for a commit:
+	// https://github.community/t/graphql-api-get-list-of-files-related-to-commit/14047/2
+	const commitInfo = await api.v3(`commits/${commits[0].sha as string}`);
+	const fileInfo = commitInfo.files.find((file: AnyObject) => [file.filename, file.previous_filename].includes(filePath));
+	if (!fileInfo) {
+		return;
+	}
+
+	const commitSha = commitInfo.sha.slice(0, 8);
+	const commitDate = commitInfo.commit.committer.date;
+	const linkToCommit = commitInfo.html_url;
+
+	url.assign({
+		route: 'commits',
+	});
+	const linkToCommitHistory = url.toString();
+
+	if (fileInfo.status === 'removed') {
+		return {
+			type: 'removed',
+			commitDetails: {
+				filePath,
+				commitSha,
+				commitDate,
+				linkToCommit,
+				linkToCommitHistory,
+				linkToLastVersion: fileInfo.blob_url,
+			},
+		};
+	}
+
+	if (fileInfo.status === 'renamed') {
+		url.assign({
+			route: 'blob',
+			branch: commitInfo.parents[0].sha,
+			filePath,
+		});
+
+		return {
+			type: 'renamed',
+			commitDetails: {
+				filePath: fileInfo.previous_filename,
+				newFilePath: fileInfo.filename,
+				commitSha,
+				commitDate,
+				linkToCommit,
+				linkToCommitHistory,
+				linkToLastVersion: url.toString(), linkToNewVersion: fileInfo.blob_url,
+			},
+		};
+	}
```

Comment:
This style isn't working, perhaps it should just return `{fileInfo, commitInfo}` and then create the DOM directly in `showHelpfulLinks`. 

One thing that will help in there is to generate the shared DOM chunk only once:

```jsx
const lastVersion = <a href={change.commitDetails.linkToLastVersion}>The file</a>;
const permalink = (
	  <a href={change.commitDetails.linkToCommit}>
		  <relative-time datetime={change.commitDetails.commitDate}/>
	  </a>
);

const commitHistory = <a href={change.commitDetails.linkToCommitHistory}>Commit history</a>;

const verb = status === 'removed'
	? deleted
	: <a href={change.commitDetails.linkToNewVersion}>moved</a>

select(…).append(
	<p className="container mt-4 text-center">
		${lastVersion} was {verb} (${permalink}) - ${commitHistory}.
	</p>		
)
```

1. 👆 no extra text before "Commit history".
2. Don't create the intermediary `change` object, my code is only a demo.